### ISSUE TITLE
Shuffle Step

### ISSUE THREAD

**Issue by brycemcanally** [pachyderm/pachyderm#3338]
**Problem:**

During the development of Pachyderm 1.8, many improvements were made to speed up the merging process, but this process still needs improvements and has some issues at high datum counts within a single job. The merging process makes many object storage requests (proportional to the number of datums) because it uses object storage as the intermediary storage for metadata. This high number of requests has caused unpredicatable object storage behavior in s3 on certain workloads because we make the requests much quicker and more bursty than the old merge which was much slower. 

**Potential Solution:**

In order to avoid the high number of object storage requests during the merge, we can shuffle the output metadata between the workers within the cluster instead of using object storage as an intermediary storage for the merge. The exact implementation of this can vary, but I will lay out a high level algorithm that seems to make sense.

- While a worker is processing datums in a chunk, it writes out the metadata that will be sent to each shard to files that correspond to each of the shards. 
- When the worker has finished processing a chunk, the metadata will be sent to the appropriate workers responsible for each shard.
- A chunk will be marked as complete when both the processing is complete and the metadata has been sent to the appropriate workers.
- The workers responsible for shards will perform the merge using this local data.

**Potential Concerns:**

_What happens if a machine that fails had a worker who was responsible for a shard and we lose the datum hashtrees?_

A way to handle this is to have the hashtrees also available in object storage in some format. In terms of first steps, we may want to have workers continue to write out datum hashtrees as they are processed because writing out the datum hashtree has not proved problematic up to this point. This would mean that we could download the hashtrees from object storage in the case that a machine with a worker responsible for a shard crashed and this would also allow us to introduce this change without needing migration. In the long run, we would probably want to batch output hashtrees and have them pre-filtered (objects correspond to shard metadata rather than full datum hashtrees).

**Comment by brycemcanally** [comment]
**Update**

This is an update on where we are in terms of the shuffle step implementation. The model that we decided to go with is a pull-based model rather than a push-based model, which was what we described in the original design. The pull-based model has much cleaner failure semantics and is a simpler implementation that meets our performance requirements. The set of operations that happen now look like this:

- While a worker is processing a chunk, it is writing the completed datum hashtrees to both object storage and a local datum cache. 
- When a worker finishes processing a chunk, it merges the datum hashtrees in the local datum cache into a hashtree that it stores in the local chunk cache (which stores merged chunks). Then, it marks the chunk as complete with its address in the chunk state.
- As chunks are completing, workers responsible for shards are requesting the chunk hashtrees from the workers that processed them, and thus have them cached locally. The requesting workers also send a shard identifier that the workers will use to filter the results to just the data relevant to the requester. If a request for a particular chunk hashtree fails, then these workers will fall back on the old model to pull and filter datum hashtrees from object storage. This fallback happens at a per-chunk granularity, so one worker going down does not cause us to fallback to object storage for the later chunks.
- After all processing is complete, the workers responsible for shards will merge the chunk hashtrees that are stored locally.

**Notes:**
- Something new added that was not really talked about in the original post is new shard claiming logic that persists across jobs. When a worker starts, it will attempt to claim one of the shards and keep it for as long as possible. The goal with this is to support caching across jobs (to avoid re-downloading data for subsequent jobs) in the future. If a worker goes down or fails to renew its claim on a shard, all of the workers without shards will attempt to acquire the newly available shard (which may include the worker who originally claimed the shard)
- In the full re-merge case (re-processed datums) each worker will download the skipped datum hashtrees in the chunks they process and merge them into the chunk tree. This a simple to reason about model, particularly in terms of the actual implementation, since the workers doing a merge will only need to have logic for determining if it is using a parent or not in the merge.

**Comment by brycemcanally** [comment]
**Discussion on whether we really want the shuffle aspect of the work we have done so far rather than better object storage batching**

We discovered a serious bug in the process of implementing the shuffle step that may have been the primary contributor to the unpredictable object storage behavior. The bug essentially was a busy loop in each worker that failed to claim a chunk/merge when they were waiting on the last set of chunks/merges to complete. This resulted with a ton of CPU usage and network activity (100% CPU usage and ~2.5 million packets per node in a 10 node cluster (20 workers / 5 mergers) with machine type of r4.xlarge in AWS on a simple 100,000 datum workload). After discovering this, we decided to think a bit more about whether we really want shuffling logic or if just more effective batching to object storage would meet our needs. We are having some trouble reproducing the issues some users have been seeing with the merge hanging, but it is worth considering whether continuing down the route of managing a shuffle makes sense from a complexity vs value standpoint. 

Pros of shuffling:
- A constant factor improvement in merge time because we remove the object storage hop to get the data to the correct merger (this will come into play more when we are merging the actual data).
- Could be noticeably less expensive in the case where there are many chunks

Pros of just batching on top of object storage:
- Less complexity

I am leaning towards keeping the shuffling right now because the constant factor improvement could be very noticeable and the complexity does not seem excessive. In terms of complexity, the main concern is future complexity that will come with moving the actual output data between workers to be merged. I think for the most part the issues we would run into there (too much data for disk) will have to be addressed even if we are using object storage for the intermediary because we need to get all the data through disk and merged. There probably will be some check pointing during the merge in those particular cases.
